polars/polars-lazy/polars-plan/src/dsl/function_expr/temporal.rs

```diff
@@ -28,22 +32,36 @@ pub(super) fn date_offset(s: Series, offset: Duration) -> PolarsResult<Series> {
                 }
             }
 
-            let out = match tz {
+            let offset_fn = offset_fn(tu);
+
+            let tz_args = match tz {
                 #[cfg(feature = "timezones")]
-                Some(ref tz) => {
-                    let offset_fn = offset_fn(tu);
-                    ca.0.try_apply(|v| offset_fn(&offset, v, tz.parse::<Tz>().ok().as_ref()))
-                }
-                _ => {
-                    let offset_fn = offset_fn(tu);
-                    ca.0.try_apply(|v| offset_fn(&offset, v, None))
+                Some(ref tz) => tz.parse::<Tz>().ok(),
+                _ => None,
+            };
+
+            let out = match offsets.len() {
+                1 => {
+                    let offset = match offsets.get(0) {
+                        Some(offset) => Duration::parse(offset),
+                        _ => Duration::parse("0"),
+                    };
+                    ca.0.try_apply(|v| offset_fn(&offset, v, tz_args.as_ref()))
                 }
+                _ => Ok(ca.0.apply_with_idx(|(idx, v)| {
+                    let offset = match offsets.get(idx) {
```

Comment:
`offsets.get` is very slow. 

This part of the code should iterate both over `ca` and `offsets` directly.

Reply:
Thanks - I had a feeling `.apply_with_idx` was probably the wrong way to go about things.

Does this mean we're not supposed to use `.apply_*()` in this case?

Are we supposed to iterate as in `ca.into_iter().zip(offsets.into_iter()).map().collect()` and create a new ChunkedArray from the result?

Reply:
Hmm.. I think this pattern will be used more often, let met write something down in the contribution guide as a cookiecutter template for this!

Reply:
@cmdlineluser can you take a look at #10013 . That should help to get you started.

Reply:
Thanks @ritchie46 - reading now.

Reply:
I'm afraid my lack of rust skills may be a problem here, hopefully this makes some bit of sense:

From what I can make out, it looks like the `compute_kernel()` examples work on arrays of the same type `T`?

In this case, I'm left with datetimes which `.downcast_iter()` into a `PrimativeArray` but the offsets end up as `Utf8Array` as they are just strings at this point.

It's a bit unclear to me how I'm supposed to run the offset strings through `Duration::parse()`.

I see there are Duration types listed for [ArrowPrimitiveType](https://docs.rs/arrow/44.0.0/arrow/datatypes/trait.ArrowPrimitiveType.html) - so I am guessing it would downcast into a `PrimativeArray` if I had a Duration type instead of just strings?